core/drivers/crypto/stm32/stm32_pka.c

```diff
@@ -0,0 +1,1740 @@
+
+	io_setbits32(base + _PKA_CR, _PKA_CR_EN);
+
+	return pka_wait_bit(base, _PKA_SR_INITOK);
+}
+
+/*
+ * Data are already loaded in PKA internal RAM,
+ * MODE is set,
+ * we start process, and wait for its end.
+ */
+static TEE_Result stm32_pka_process(const vaddr_t base)
+{
+	io_setbits32(base + _PKA_CR, _PKA_CR_START);
+
+	return pka_wait_bit(base, _PKA_IT_PROCEND);
+}
+
+/**
+ * Read ECC operand from PKA RAM
+ *
+ * PKA reads u64 words, for each u64 LSB is bit 0, MSB is bit 63.
+ * We read @eo_nbw (ECC operand Size) u64. The value of @eo_nbw depends
+ * on the chosen prime modulus length in bits.
+ *
+ * First less significant u64 is read from lowest address.
+ * Most significant u64 from higher address.
+ * Last u64 is expect to be equal to 0x0.
+ *
+ * This function manage endianness (as bswap64 do) and padding incomplete u64
+ * with 0 (if data is not a u64 multiple).
+ *
+ * @addr: PKA_RAM address to read from to the buffer @data.
+ * @data: will be a BYTE list with most significant bytes first.
+ * @data_size: [in] @data size,
+ *             [out] nb of bytes in data.
+ * @eo_nbw: is ECC Operand size in 64bits word (including the extra 0)
+ *                (note it depends on the prime modulus length, not the data
+ *                size).
+ * @return TEE_SUCCESS if OK.
+ *         TEE_ERROR_SECURITY if the last u64 word is not 0.
+ *         TEE_ERROR_BAD_PARAMETERS if data_size and eo_nbw are inconsistent,
+ *         i.e. data doesn't fit in defined eo_nbw, or eo_nbw bigger than
+ *         hardware limit, or if [in]data_size is too small to get the data.
+ */
+static TEE_Result read_eo_data(const vaddr_t addr, uint8_t *data,
+			       const unsigned int data_size,
+			       const unsigned int eo_nbw)
+{
+	uint32_t word_index = U(0);
+	int data_index = (int)data_size - 1;
+	uint64_t tmp = ULL(0);
+
+	if (eo_nbw < OP_NBW_FROM_SIZE(data_size) || eo_nbw > MAX_EO_NBW)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/* Fill value */
+	for (word_index = U(0); word_index < eo_nbw - 1; word_index++) {
+		/* Index in the tmp U64 word */
+		unsigned int i = U(0);
+
+		tmp = io_read64(addr + word_index * sizeof(tmp));
+
+		while ((i < sizeof(tmp)) && (data_index >= 0)) {
+			data[data_index] = tmp & 0xFF;
+			tmp = tmp >> INT8_LEN;
+
+			/* Move byte index in current (u64)tmp */
+			i++;
+
+			/* Move to next most significant byte */
+			data_index--;
+		}
+	}
+
+	/* The last u64 should be 0 */
+	tmp = io_read64(addr + word_index * sizeof(tmp));
+	if (tmp)
+		return TEE_ERROR_SECURITY;
+
+	return TEE_SUCCESS;
+}
+
+/**
+ * Write ECC operand to PKA RAM.
+ *
+ * PKA expects to write u64 word, each u64 are: the least significant bit
+ * is bit 0; the most significant bit is bit 63.
+ * We write eo_nbw (ECC operand Size) u64, value that depends on the
+ * chosen prime modulus length in bits.
+ *
+ * First less significant u64 is written to low address.
+ * Most significant u64 to higher address.
+ * And at last address we write a u64(0x0).
```

Comment:
"First less significant u64 is written to low**est** address."
"Most significant u64 **words are written** to the higher address**es**." (or remove that sentence, implicitly said by line above)